### ISSUE TITLE
codgen Register Coalescing crash

### ISSUE THREAD

**Issue by OskarLinde** [llvm-mos/llvm-mos#257]
When compiling with optimizations `-O`, `-Oz`, etc the following (relatively minimal) sample crashes.

```c
void func(signed char x0, signed char x1) {
    signed char dx,sx;
    short ci = 0;

    if (x0 < x1) { dx = x1 - x0; sx =  1; }
    else         { dx = x0 - x1; sx = -1; }

    while(1) {
        if (dx >= ci) { ci += sx; }
    }
}

int main() {
    #define RAM       ((volatile unsigned char*)0x0400)
    func(RAM[0], RAM[1]);
}
```

e.g. `mos-c64-clang -O test.c`
`clang version 16.0.0 (https://www.github.com/llvm-mos/llvm-mos.git 8917a1fc71299ab8c0fa4db56c3c03c3184beca4)`

```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /Users/oskar/llvm-mos/bin/ld.lld /var/folders/yx/wx9gtv657vq3xlkt5hk_f1w40000gn/T/test_min-e9fc8c.o -plugin-opt=O1 -mllvm -force-precise-rotation-cost -mllvm -jump-inst-cost=6 -mllvm -force-loop-cold-block -mllvm -phi-node-folding-threshold=0 -mllvm -two-entry-phi-node-folding-threshold=0 -mllvm -align-large-globals=false -mllvm -disable-spill-hoist -mllvm -zp-avail=110 --gc-sections --sort-section=alignment -L/Users/oskar/llvm-mos/bin/../mos-platform/c64/lib -L/Users/oskar/llvm-mos/bin/../mos-platform/commodore/lib -L/Users/oskar/llvm-mos/bin/../mos-platform/common/lib -l:crt0.o -lcrt0 -lcrt -lc -Tlink.ld -o a.out
1.	Running pass 'Function Pass Manager' on module 'ld-temp.o'.
2.	Running pass 'Simple Register Coalescing' on function '@main'
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  lld                      0x0000000102ccc11f llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 43
1  lld                      0x0000000102ccc504 SignalHandler(int) + 168
2  libsystem_platform.dylib 0x00007ff81d8d2dfd _sigtramp + 29
3  libsystem_platform.dylib 0x0000000200000000 _sigtramp + 18446603378675077664
4  lld                      0x0000000103415db7 (anonymous namespace)::RegisterCoalescer::copyCoalesceWorkList(llvm::MutableArrayRef<llvm::MachineInstr*>) + 139
5  lld                      0x0000000103413bf0 (anonymous namespace)::RegisterCoalescer::runOnMachineFunction(llvm::MachineFunction&) + 3462
6  lld                      0x000000010334923c llvm::MachineFunctionPass::runOnFunction(llvm::Function&) + 560
7  lld                      0x0000000103fea55a llvm::FPPassManager::runOnFunction(llvm::Function&) + 506
8  lld                      0x0000000103fededa llvm::FPPassManager::runOnModule(llvm::Module&) + 52
9  lld                      0x0000000103fea97b llvm::legacy::PassManagerImpl::run(llvm::Module&) + 649
10 lld                      0x0000000103259c73 codegen(llvm::lto::Config const&, llvm::TargetMachine*, std::__1::function<llvm::Expected<std::__1::unique_ptr<llvm::CachedFileStream, std::__1::default_delete<llvm::CachedFileStream>>> (unsigned int)>, unsigned int, llvm::Module&, llvm::ModuleSummaryIndex const&) + 1294
11 lld                      0x0000000103259230 llvm::lto::backend(llvm::lto::Config const&, std::__1::function<llvm::Expected<std::__1::unique_ptr<llvm::CachedFileStream, std::__1::default_delete<llvm::CachedFileStream>>> (unsigned int)>, unsigned int, llvm::Module&, llvm::ModuleSummaryIndex&) + 302
12 lld                      0x000000010324ebb3 llvm::lto::LTO::runRegularLTO(std::__1::function<llvm::Expected<std::__1::unique_ptr<llvm::CachedFileStream, std::__1::default_delete<llvm::CachedFileStream>>> (unsigned int)>) + 1215
13 lld                      0x000000010324e4ab llvm::lto::LTO::run(std::__1::function<llvm::Expected<std::__1::unique_ptr<llvm::CachedFileStream, std::__1::default_delete<llvm::CachedFileStream>>> (unsigned int)>, std::__1::function<llvm::Expected<std::__1::function<llvm::Expected<std::__1::unique_ptr<llvm::CachedFileStream, std::__1::default_delete<llvm::CachedFileStream>>> (unsigned int)>> (unsigned int, llvm::StringRef)>) + 569
14 lld                      0x0000000102dd0926 lld::elf::BitcodeCompiler::compile() + 558
15 lld                      0x0000000102d809c7 void lld::elf::LinkerDriver::compileBitcodeFiles<llvm::object::ELFType<(llvm::support::endianness)1, false>>(bool) + 177
16 lld                      0x0000000102d7c40c lld::elf::LinkerDriver::link(llvm::opt::InputArgList&) + 4678
17 lld                      0x0000000102d73fb8 lld::elf::LinkerDriver::linkerMain(llvm::ArrayRef<char const*>) + 3740
18 lld                      0x0000000102d73088 lld::elf::link(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, bool, bool) + 1083
19 lld                      0x0000000102c05e2c lldMain(int, char const**, llvm::raw_ostream&, llvm::raw_ostream&, bool) + 1173
20 lld                      0x0000000102c058d4 lld_main(int, char**) + 349
21 dyld                     0x000000020526f52e start + 462
22 dyld                     0x000000020526a000 start + 18446744073709530272
clang-16: error: unable to execute command: Segmentation fault: 11
clang-16: error: ld.lld command failed due to signal (use -v to see invocation)
```

Other permutations of the sample causes other stack traces, e.g.

```c
void func(signed char x0) {
    signed char dx,sx,x1;  // Note x1 uninitialized. x1=1 restores original crash
    short ci = 0;

    if (x0 < x1) { dx = x1 - x0; sx =  1; }
    else         { dx = x0 - x1; sx = -1; }

    while(1) {
        if (dx >= ci) { ci += sx; }
    }
}

int main() {
    func(((volatile unsigned char*)0x0400)[0]);
}
```

`mos-c64-clang -O test.c`

produces a few different (random) crashes

```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /Users/oskar/llvm-mos/bin/ld.lld /var/folders/yx/wx9gtv657vq3xlkt5hk_f1w40000gn/T/test_min-f64267.o -plugin-opt=O1 -mllvm -force-precise-rotation-cost -mllvm -jump-inst-cost=6 -mllvm -force-loop-cold-block -mllvm -phi-node-folding-threshold=0 -mllvm -two-entry-phi-node-folding-threshold=0 -mllvm -align-large-globals=false -mllvm -disable-spill-hoist -mllvm -zp-avail=110 --gc-sections --sort-section=alignment -L/Users/oskar/llvm-mos/bin/../mos-platform/c64/lib -L/Users/oskar/llvm-mos/bin/../mos-platform/commodore/lib -L/Users/oskar/llvm-mos/bin/../mos-platform/common/lib -l:crt0.o -lcrt0 -lcrt -lc -Tlink.ld -o a.out
1.	Running pass 'Function Pass Manager' on module 'ld-temp.o'.
2.	Running pass 'Optimize MOS Increment/Decrement PHI pattern' on function '@main'
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  lld                      0x000000010211411f llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 43
1  lld                      0x0000000102114504 SignalHandler(int) + 168
2  libsystem_platform.dylib 0x00007ff81d8d2dfd _sigtramp + 29
3  libsystem_platform.dylib 0x000000030a7d9470 _sigtramp + 18446603383146047120
4  libsystem_malloc.dylib   0x00007ff81d6f4abb _malloc_zone_malloc + 125
5  libsystem_malloc.dylib   0x00006000033a00a0 _malloc_zone_malloc + 18446708923257501282
clang-16: error: unable to execute command: Segmentation fault: 11
clang-16: error: ld.lld command failed due to signal (use -v to see invocation)
```

```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
clang-16: error: unable to execute command: Bus error: 10
clang-16: error: ld.lld command failed due to signal (use -v to see invocation)
```

**Comment by mysterymath** [comment]
If you don't mind my asking, how did you find this one? After reproducing, it looks like the issue only occurs in a fairly specific scenario.

**Comment by OskarLinde** [comment]
Thanks for the fast resolution! Can confirm that the diff worked.

Stumbled on it by pure coincidence. Happy to hear that it was a rare corner. Here's a longer sample – a simple Bresenham line algorithm:

```c
typedef unsigned char  uint8_t;
typedef signed char    int8_t;
typedef unsigned short uint16_t;

const uint8_t W = 40;

#define SCREEN_RAM       ((volatile unsigned char*)0x0400)

void plot(uint16_t ci) { SCREEN_RAM[ci] = 160; }

void bresenham(int8_t x0, int8_t y0, int8_t x1, int8_t y1) {
    int8_t dx,dy,sx,sy;

    if (x0 < x1) { dx = x1 - x0; sx =  1; }
    else         { dx = x0 - x1; sx = -1; }
    if (y0 < y1) { dy = y0 - y1; sy =  W; } // note dy is negative
    else         { dy = y1 - y0; sy = -W; }

    int8_t error = dx + dy;

    uint16_t ci = x0 + y0 * W;
    uint16_t ct = x1 + y1 * W;

    while(1) {
        plot(ci);
        if (ci == ct) return;
        int8_t e2 = 2 * error;

        if (e2 >= dy) { error += dy; ci += sx; }
        if (e2 <= dx) { error += dx; ci += sy; }
    }
}

int main() {
    bresenham(3,20,37,3);
}
```

**Comment by mysterymath** [comment]
Ah, that makes sense. The actual issue was that it noticed that ci is incremented by sx, but that sx is either 1 or -1 based on an if. I wrote a specific optimization to turn this into an increment and decrement in the then and else branches of the if.

This code wasn't well enough tested, and this case slipped through. Here, it's not safe to do that transformation, since ci isn't known until after the if else structure above. The transform added increments and decrements of ci to those to then and else blocks, and the compiler crashed, since ci wasn't defined there. I just added a check that the definition of ci is available at the proposed increment and decrement sites, which disables the optimization in this case, but allows it to proceed in others.

Thanks for the report!
